BigQuery slot purchase fails in `tempdev`

https://gitlab.temp.gi.ucsc.edu/ucsc/azul/-/jobs/68

```
2024-05-15 03:42:40,522    INFO MainThread __main__: Queuing notifications for reindexing ...
2024-05-15 03:42:40,533    INFO MainThread azul.deployment: Allocated new Boto3 client for 'secretsmanager' with ID 140436782882640
2024-05-15 03:42:41,100    INFO MainThread azul.bigquery_reservation: Purchasing 100 BigQuery slots in location 'us-central1'
Traceback (most recent call last):
  File "/build/.venv/lib/python3.11/site-packages/google/api_core/grpc_helpers.py", line 75, in error_remapped_callable
    return callable_(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/build/.venv/lib/python3.11/site-packages/grpc/_channel.py", line 1181, in __call__
    return _end_unary_response_blocking(state, call, False, None)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/build/.venv/lib/python3.11/site-packages/grpc/_channel.py", line 1006, in _end_unary_response_blocking
    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.FAILED_PRECONDITION
	details = "The specified flat rate commitment plan is beyond end of sale and no longer available. Please use an Editions commitment plan."
	debug_error_string = "UNKNOWN:Error received from peer ipv4:172.253.115.95:443 {created_time:"2024-05-15T03:42:41.382229115+00:00", grpc_status:9, grpc_message:"The specified flat rate commitment plan is beyond end of sale and no longer available. Please use an Editions commitment plan."}"
>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/builds/ucsc/azul/scripts/reindex.py", line 206, in <module>
    main(sys.argv[1:])
  File "/builds/ucsc/azul/scripts/reindex.py", line 189, in main
    reservation.activate()
  File "/builds/ucsc/azul/src/azul/bigquery_reservation.py", line 123, in activate
    self._ensure_capacity_commitment()
  File "/builds/ucsc/azul/src/azul/bigquery_reservation.py", line 142, in _ensure_capacity_commitment
    self._create_capacity_commitment(self.slots)
  File "/builds/ucsc/azul/src/azul/bigquery_reservation.py", line 192, in _create_capacity_commitment
    commitment = self._client.create_capacity_commitment(capacity_commitment=commitment,
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/build/.venv/lib/python3.11/site-packages/google/cloud/bigquery_reservation_v1/services/reservation_service/client.py", line 1206, in create_capacity_commitment
    response = rpc(
               ^^^^
  File "/build/.venv/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py", line 131, in __call__
    return wrapped_func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/build/.venv/lib/python3.11/site-packages/google/api_core/timeout.py", line 120, in func_with_timeout
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/build/.venv/lib/python3.11/site-packages/google/api_core/grpc_helpers.py", line 77, in error_remapped_callable
    raise exceptions.from_grpc_error(exc) from exc
google.api_core.exceptions.FailedPrecondition: 400 The specified flat rate commitment plan is beyond end of sale and no longer available. Please use an Editions commitment plan.
make: *** [Makefile:128: reindex] Error 1
```

--- a/src/azul/bigquery_reservation.py
+++ b/src/azul/bigquery_reservation.py
@@ -7,6 +7,7 @@ from typing import (
 from google.cloud.bigquery_reservation_v1 import (
     Assignment,
     CapacityCommitment,
+    Edition,
     Reservation,
     ReservationServiceClient,
 )
@@ -123,9 +124,9 @@ class BigQueryReservation:
         if not self.dry_run:
             if not self.is_active:
                 raise RuntimeError('Failed to activate slots')
-            if self.reservation.slot_capacity < self.slots:
+            if self.reservation.autoscale.max_slots < self.slots:
                 raise RuntimeError('Failed to acquire enough slots',
-                                   self.reservation.slot_capacity,
+                                   self.reservation.autoscale.max_slots,
                                    self.slots)
 
     def _create_reservation(self) -> None:
@@ -134,22 +135,23 @@ class BigQueryReservation:
         """
         self._refresh('reservation')
         if self.reservation is None:
-            reservation = Reservation(dict(slot_capacity=self.slots,
-                                           ignore_idle_slots=False))
+            reservation = Reservation(dict(edition=Edition.STANDARD,
+                                           autoscale=Reservation.Autoscale(dict(max_slots=self.slots)),
+                                           ignore_idle_slots=True))
             if self.dry_run:
                 log.info('Would reserve %d BigQuery slots in location %r, reservation ID: %r',
-                         reservation.slot_capacity, self.location, self._reservation_id)
+                         reservation.autoscale.max_slots, self.location, self._reservation_id)
             else:
                 log.info('Reserving %d BigQuery slots in location %r, reservation ID: %r',
-                         reservation.slot_capacity, self.location, self._reservation_id)
+                         reservation.autoscale.max_slots, self.location, self._reservation_id)
                 reservation = self._client.create_reservation(reservation=reservation,
                                                               reservation_id=self._reservation_id,
                                                               parent=self._reservation_parent_path)
                 log.info('Reserved %d BigQuery slots in location %r, reservation name: %r',
-                         reservation.slot_capacity, self.location, reservation.name)
+                         reservation.autoscale.max_slots, self.location, reservation.name)
                 self.reservation = reservation
         else:
-            current_capacity = self.reservation.slot_capacity
+            current_capacity = self.reservation.autoscale.max_slots
             log.info('Reservation with capacity %d already created in location %r',
                      current_capacity, self.location)
             if current_capacity < self.slots:
@@ -158,12 +160,12 @@ class BigQueryReservation:
                     log.info('Would increase reservation capacity to %d', self.slots)
                 else:
                     log.info('Increasing reservation capacity to %d', self.slots)
-                    self.reservation.slot_capacity = self.slots
+                    self.reservation.autoscale.max_slots = self.slots
                     self.reservation = self._client.update_reservation(
                         reservation=self.reservation,
-                        update_mask='slotCapacity'
+                        update_mask='autoscale'
                     )
-                    log.info('Reservation now has capacity %d', self.reservation.slot_capacity)
+                    log.info('Reservation now has capacity %d', self.reservation.autoscale.max_slots)
 
     def _assign_slots(self) -> None:
         """

--- a/src/azul/bigquery_reservation.py
+++ b/src/azul/bigquery_reservation.py
@@ -1,5 +1,4 @@
 import logging
-import time
 from typing import (
     Optional,
     Union,
@@ -48,7 +47,6 @@ class BigQueryReservation:
         'assignment': '/reservations/-'
     }
 
-    capacity_commitment: Optional[CapacityCommitment]
     reservation: Optional[Reservation]
     assignment: Optional[Assignment]
     location: str
@@ -99,7 +97,6 @@ class BigQueryReservation:
     @property
     def is_active(self) -> Optional[bool]:
         resource_statuses = {
-            self.capacity_commitment is not None,
             self.reservation is not None,
             self.assignment is not None
         }
@@ -120,116 +117,16 @@ class BigQueryReservation:
             return self.reservation.update_time.timestamp()
 
     def activate(self) -> None:
-        self._ensure_capacity_commitment()
         self._create_reservation()
         self._assign_slots()
         self.refresh()
         if not self.dry_run:
             if not self.is_active:
                 raise RuntimeError('Failed to activate slots')
-            if self.capacity_commitment.slot_count < self.slots:
+            if self.reservation.slot_capacity < self.slots:
                 raise RuntimeError('Failed to acquire enough slots',
-                                   self.capacity_commitment.slot_count,
+                                   self.reservation.slot_capacity,
                                    self.slots)
-
-    def _ensure_capacity_commitment(self) -> None:
-        """
-        Idempotently purchase capacity commitment, ensuring that the number of
-        reserved slots is sufficient for the current configuration.
-        """
-        self._refresh('capacity_commitment')
-        if self.capacity_commitment is None:
-            self._create_capacity_commitment(self.slots)
-        else:
-            current_capacity = self.capacity_commitment.slot_count
-            log.info('Slot commitment with %d slots already purchased in location %r',
-                     current_capacity, self.location)
-            if current_capacity < self.slots:
-                deficit = self.slots - current_capacity
-                log.info('Slot deficit is %d', deficit)
-                # To increase the capacity of an existing commitment, we must
-                # create a new commitment and then merge it with the old one.
-                old_commitment = self.capacity_commitment
-                self._create_capacity_commitment(deficit)
-                new_commitment = self.capacity_commitment
-                if self.dry_run:
-                    assert old_commitment is new_commitment
-                    log.info('Would merge capacity commitments in location %r', self.location)
-                else:
-                    assert old_commitment.name != new_commitment.name
-                    log.info('Merging capacity commitments {%r, %r}',
-                             old_commitment.name, new_commitment.name)
-
-                    def extract_id(commitment: CapacityCommitment) -> str:
-                        parent, resource_type, id = commitment.name.rsplit('/', 2)
-                        assert parent == self._reservation_parent_path
-                        assert resource_type == 'capacityCommitments'
-                        return id
-
-                    self.capacity_commitment = self._client.merge_capacity_commitments(
-                        parent=self._reservation_parent_path,
-                        capacity_commitment_ids=list(map(extract_id, [old_commitment, new_commitment]))
-                    )
-                    log.info('Merged capacity commitments {%r, %r} to create %r with %d total slots',
-                             old_commitment.name, new_commitment.name, self.capacity_commitment.name,
-                             self.capacity_commitment.slot_count)
-            elif current_capacity > self.slots:
-                log.warning('Existing slot commitment has more slots than requested; '
-                            'excessive costs may be incurred. If a smaller capacity '
-                            'is desired, it must be reduced manually.')
-            else:
-                log.info('No purchase necessary')
-
-    def _create_capacity_commitment(self, slots: int) -> None:
-        commitment = CapacityCommitment(dict(slot_count=slots,
-                                             plan=CapacityCommitment.CommitmentPlan.FLEX))
-        if self.dry_run:
-            log.info('Would purchase %d BigQuery slots in location %r',
-                     commitment.slot_count, self.location)
-        else:
-            log.info('Purchasing %d BigQuery slots in location %r',
-                     commitment.slot_count, self.location)
-            commitment = self._client.create_capacity_commitment(capacity_commitment=commitment,
-                                                                 parent=self._reservation_parent_path)
-            log.info('Purchased %d BigQuery slots in location %r, commitment name: %r',
-                     commitment.slot_count, self.location, commitment.name)
-            # Record state before waiting for activation so that we can delete it on failure
-            self.capacity_commitment = commitment
-            self.capacity_commitment = self._await_active_commitment(commitment)
-
-    def _await_active_commitment(self,
-                                 commitment: CapacityCommitment
-                                 ) -> CapacityCommitment:
-        """
-        Poll for a minute or until commitment is active. Fail gracefully if we
-        are unable to get commitment. See Google's docs for more info:
-        https://cloud.google.com/bigquery/docs/reservations-tasks#purchased_slots_are_pending
-        """
-        start = time.time()
-        deadline = start + 60
-        now = start
-        while commitment.state != commitment.State.ACTIVE:
-            if now > deadline:
-                self.deactivate()
-                log.error('Commitment %r in state %r after %.3fs seconds. '
-                          'Commitment was deleted. Try again later.',
-                          commitment.name, commitment.state.name, now - start)
-                raise RuntimeError('Slot commitment not active in time')
-            elif commitment.state == commitment.State.FAILED:
-                self.deactivate()
-                raise RuntimeError('Slot commitment failed to activate',
-                                   commitment.failure_status)
-            elif commitment.state == commitment.State.PENDING:
-                log.info('Commitment %r pending. Trying again in 10 seconds...',
-                         commitment.name)
-                time.sleep(10)
-                commitment = self._client.get_capacity_commitment(name=commitment.name)
-                now = time.time()
-            else:
-                assert False, commitment.state
-        log.info('Commitment %r is active after %.3fs seconds',
-                 commitment.name, now - start)
-        return commitment
 
     def _create_reservation(self) -> None:
         """

